pkg/sql/copy/copy_test.go

```diff
@@ -0,0 +1,301 @@
+	l_orderkey      INT8 NOT NULL,
+	l_partkey       INT8 NOT NULL,
+	l_suppkey       INT8 NOT NULL,
+	l_linenumber    INT8 NOT NULL,
+	l_quantity      DECIMAL(15,2) NOT NULL,
+	l_extendedprice DECIMAL(15,2) NOT NULL,
+	l_discount      DECIMAL(15,2) NOT NULL,
+	l_tax           DECIMAL(15,2) NOT NULL,
+	l_returnflag    CHAR(1) NOT NULL,
+	l_linestatus    CHAR(1) NOT NULL,
+	l_shipdate      DATE NOT NULL,
+	l_commitdate    DATE NOT NULL,
+	l_receiptdate   DATE NOT NULL,
+	l_shipinstruct  CHAR(25) NOT NULL,
+	l_shipmode      CHAR(10) NOT NULL,
+	l_comment       VARCHAR(44) NOT NULL,
+	PRIMARY KEY     (l_orderkey, l_linenumber),
+	INDEX l_ok      (l_orderkey ASC),
+	INDEX l_pk      (l_partkey ASC),
+	INDEX l_sk      (l_suppkey ASC),
+	INDEX l_sd      (l_shipdate ASC),
+	INDEX l_cd      (l_commitdate ASC),
+	INDEX l_rd      (l_receiptdate ASC),
+	INDEX l_pk_sk   (l_partkey ASC, l_suppkey ASC),
+	INDEX l_sk_pk   (l_suppkey ASC, l_partkey ASC)
+)`
+
+const csvData = `%d|155190|7706|1|17|21168.23|0.04|0.02|N|O|1996-03-13|1996-02-12|1996-03-22|DELIVER IN PERSON|TRUCK|egular courts above the`
+
+func TestCopy(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	ctx := context.Background()
+
+	s, _, _ := serverutils.StartServer(t, base.TestServerArgs{
+		Settings: cluster.MakeTestingClusterSettings(),
+	})
+	defer s.Stopper().Stop(ctx)
+
+	url, cleanup := sqlutils.PGUrl(t, s.ServingSQLAddr(), t.Name(), url.User(username.RootUser))
+	defer cleanup()
+	var sqlConnCtx clisqlclient.Context
+	conn := sqlConnCtx.MakeSQLConn(ioutil.Discard, ioutil.Discard, url.String())
+
+	testCopy := func(t *testing.T, d *datadriven.TestData) string {
+		switch d.Cmd {
+		case "exec-ddl":
+			err := conn.Exec(ctx, d.Input)
+			if err != nil {
+				require.NoError(t, err, "%s: %s", d.Pos, d.Cmd)
+			}
+			return ""
+		case "copy", "copy-error":
+			lines := strings.Split(d.Input, "\n")
+			stmt := lines[0]
+			data := strings.Join(lines[1:], "\n")
+			rows, err := conn.GetDriverConn().CopyFrom(ctx, strings.NewReader(data), stmt)
+			if d.Cmd == "copy" {
+				require.NoError(t, err, "%s\n%s\n", d.Cmd, d.Input)
+				require.Equal(t, int(rows), len(lines)-1, "Not all rows were inserted")
+			} else {
+				return err.Error()
+			}
+			return fmt.Sprintf("%d", rows)
+		case "query":
+			rows, err := conn.Query(ctx, d.Input)
+			require.NoError(t, err)
+
+			vals := make([]driver.Value, len(rows.Columns()))
+			var results string
+			for {
+				if err := rows.Next(vals); err == io.EOF {
+					break
+				} else if err != nil {
+					require.NoError(t, err)
+				}
+				for i, v := range vals {
+					if i > 0 {
+						results += "|"
+					}
+					results += fmt.Sprintf("%v", v)
+				}
+				results += "\n"
+			}
+			err = rows.Close()
+			require.NoError(t, err)
+			return results
+		default:
+			return fmt.Sprintf("unknown command: %s\n", d.Cmd)
+		}
+
+	}
+	datadriven.RunTest(t, testutils.TestDataPath(t, "copyfrom"), testCopy)
```

Comment:
instead of `copyfrom` being a file, should it be a directory? then this will automatically test every file in that directory

Reply:
I was just following convention for when testdata only had one file, if we decide to add more we can throw in a dir walk here but seems like overkill ATM.